sound/soc/sof/topology.c

```diff
@@ -2648,6 +2648,13 @@ static int sof_set_dai_config(struct snd_sof_dev *sdev, u32 size,
 			continue;
 
 		if (strcmp(link->name, dai->name) == 0) {
+			/*
+			 * the same dai config will be applied to all DAIs in
+			 * the same dai link. We have to ensure that the ipc
+			 * dai config's dai_index match to the component's
+			 * dai_index.
+			 */
+			config->dai_index = dai->comp_dai.dai_index;
```

Comment:
@bardliao Can you maybe give examples in topology m4 how this looks like? It's starts to sound messy if we have dai index values in topology, but for some DAIs we override the values at runtime.

Reply:
> @bardliao Can you maybe give examples in topology m4 how this looks like? It's starts to sound messy if we have dai index values in topology, but for some DAIs we override the values at runtime.

@kv2019i config->dai_index will be the same as dai->comp_dai.dai_index if we find exactly one dai in the loop. The reason why we need to overwrite config->dai_index is that we have only one dai config in the topology file and we want to share the config to all of the dais which connect to the same link.
Please see [sof-icl-rt711-rt1308-rt715-hdmi.m4](https://github.com/thesofproject/sof/blob/master/tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4). ALH0x102 and ALH0x202 are both connected to SDW1-Playback link. However, we only config [ALH_CONFIG(ALH_CONFIG_DATA(ALH, 0x102, 48000, 2))](https://github.com/thesofproject/sof/blob/master/tools/topology/sof-icl-rt711-rt1308-rt715-hdmi.m4#L219) for ALH0x102. So we can get ipc config with dai_index 0x102 in [sof_link_alh_load](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/topology.c#L3114), but no ipc config with dai_index 0x202. However, we do need ipc config with dai_index 0x202. So we made an assumption that all dais which connect to the same link have the same dai config and that's why we need to set different dai_index to each ipc message. It is more like copy a ipc config to all dais which are connected to the same link. Some discussion is on https://github.com/thesofproject/linux/pull/1976

Reply:
@kv2019i this is not new. Take a look at sof_link_hda_process(), we do that for HDA DAIs already

Reply:
Thanks both @bardliao @ranj063 , that was very informative. As this was already discussed in #1976, no need to repeat here. 

Food for thought for future work: from perspective of modelling graphs, this kind of distribution of assumptions can be very problematic. A standalone tool to create topologies (or validate existing ones), would have to copy all these additional rules to do its job. E.g. to understand that there is a valid ALK_CONFIG for 0x202, you just have to know that 0x102 config will be copied. I'd argue our graph description language should be able to express information like this. But yeah, water under the bridge now.